File: DS3Lib/Sources/DS3Lib/Constants/URLs.swift
import Foundation

/// Instance-based URL builder that derives all Cubbit API URLs from a coordinator base URL.
/// Replaces the former static enum to support custom coordinator URLs for multi-tenant deployments.
public final class CubbitAPIURLs: Sendable {
    /// The default Cubbit coordinator URL.
    public static let defaultCoordinatorURL = "https://api.eu00wi.cubbit.services"

    /// The coordinator base URL used to derive all service URLs.
    public let coordinatorURL: String

    /// Creates a new CubbitAPIURLs instance.
    /// - Parameter coordinatorURL: The coordinator base URL. Defaults to the standard Cubbit coordinator.
    public init(coordinatorURL: String = CubbitAPIURLs.defaultCoordinatorURL) {
        var url = coordinatorURL
        while url.hasSuffix("/") {
            url = String(url.dropLast())
        }
        self.coordinatorURL = url
    }

    // MARK: - IAM URLs

    /// IAM service base URL
    public var iamBaseURL: String {
        "\(coordinatorURL)/iam/v1"
    }
    /// Auth base URL
    public var authBaseURL: String {
        "\(iamBaseURL)/auth"
    }
    /// Sign-in URL
    public var signinURL: String {
        "\(authBaseURL)/signin"
    }
    /// Challenge URL for authentication
    public var challengeURL: String {
        "\(signinURL)/challenge"
    }
    /// Token refresh URL
    public var tokenRefreshURL: String {
        "\(authBaseURL)/refresh/access"
    }
    /// Forge access JWT URL
    public var forgeAccessJWTURL: String {
        "\(authBaseURL)/forge/access"
    }
    /// Accounts "me" URL
    public var accountsMeURL: String {
        "\(iamBaseURL)/accounts/me"
    }

    // MARK: - Composer Hub URLs

    /// Composer Hub base URL
    public var composerHubBaseURL: String {
        "\(coordinatorURL)/composer-hub/v1"
    }
    /// Projects URL
    public var projectsURL: String {
        "\(composerHubBaseURL)/projects"
    }
    /// Tenants URL
    public var tenantsURL: String {
        "\(composerHubBaseURL)/tenants"
    }

    // MARK: - Keyvault URLs

    /// Keyvault base URL
    public var keyvaultBaseURL: String {
        "\(coordinatorURL)/keyvault/api/v3"
    }
    /// Keys URL (for create, get, delete operations)
    public var keysURL: String {
        "\(keyvaultBaseURL)/keys"
    }
}

/// Cubbit web console related URLs.
public enum ConsoleURLs {
    public static let baseURL = "https://console.cubbit.eu"
    public static let signupURL = "\(ConsoleURLs.baseURL)/signup"
    public static let recoveryURL = "\(ConsoleURLs.baseURL)/recovery"
    public static let workspaceURL = "\(ConsoleURLs.baseURL)/workspace"
    public static let projectsURL = "\(ConsoleURLs.workspaceURL)/projects"
    public static let profileURL = "\(ConsoleURLs.workspaceURL)/profile"
}

/// Cubbit docs related URLs.
public enum DocsURLs {
    public static let baseURL = "https://docs.cubbit.io"
}

/// Cubbit help center related URLs.
public enum HelpURLs {
    public static let baseURL = "https://help.cubbit.io"
}

File: core/ds3-http/src/urls.rs
//! Cubbit API URL generation from a coordinator base URL.
//!
//! Ports the Swift `CubbitAPIURLs` class. All endpoint URLs are derived from
//! a single coordinator base URL via string concatenation.

/// Generates all Cubbit API endpoint URLs from a coordinator base URL.
///
/// Default coordinator: `https://api.eu00wi.cubbit.services`
#[derive(Debug, Clone)]
pub struct CubbitAPIURLs {
    coordinator_url: String,
}

impl CubbitAPIURLs {
    /// Default Cubbit coordinator URL.
    pub const DEFAULT_COORDINATOR_URL: &str = "https://api.eu00wi.cubbit.services";

    /// Creates a new URL generator, stripping trailing slashes from the coordinator URL.
    pub fn new(coordinator_url: &str) -> Self {
        let mut url = coordinator_url.to_string();
        while url.ends_with('/') {
            url.pop();
        }
        Self {
            coordinator_url: url,
        }
    }

    /// Creates a URL generator with the default coordinator.
    pub fn default_coordinator() -> Self {
        Self::new(Self::DEFAULT_COORDINATOR_URL)
    }

    fn iam_base_url(&self) -> String {
        format!("{}/iam/v1", self.coordinator_url)
    }

    fn auth_base_url(&self) -> String {
        format!("{}/auth", self.iam_base_url())
    }

    /// Returns the challenge endpoint URL.
    pub fn challenge_url(&self) -> String {
        format!("{}/signin/challenge", self.auth_base_url())
    }

    /// Returns the signin endpoint URL.
    pub fn signin_url(&self) -> String {
        format!("{}/signin", self.auth_base_url())
    }

    /// Returns the token refresh endpoint URL.
    pub fn token_refresh_url(&self) -> String {
        format!("{}/refresh/access", self.auth_base_url())
    }

    /// Returns the forge access JWT endpoint URL.
    pub fn forge_access_jwt_url(&self) -> String {
        format!("{}/forge/access", self.auth_base_url())
    }

    /// Returns the accounts/me endpoint URL.
    pub fn accounts_me_url(&self) -> String {
        format!("{}/accounts/me", self.iam_base_url())
    }

    fn composer_hub_base_url(&self) -> String {
        format!("{}/composer-hub/v1", self.coordinator_url)
    }

    /// Returns the projects endpoint URL.
    pub fn projects_url(&self) -> String {
        format!("{}/projects", self.composer_hub_base_url())
    }

    /// Returns the tenants endpoint URL.
    pub fn tenants_url(&self) -> String {
        format!("{}/tenants", self.composer_hub_base_url())
    }

    fn keyvault_base_url(&self) -> String {
        format!("{}/keyvault/api/v3", self.coordinator_url)
    }

    /// Returns the keys endpoint URL.
    pub fn keys_url(&self) -> String {
        format!("{}/keys", self.keyvault_base_url())
    }
}
